File: ios/CBMCapture/Core/Imaging/ImageTransform.swift
import CoreGraphics
import Foundation

/// A pinhole camera in the pixel frame of a specific image.
///
/// The `width`/`height` fields are not decoration: they are what make this type meaningful.
/// An intrinsics matrix without the frame it belongs to is the defect described in section 3
/// of the calibration note, so the two never travel separately here.
struct PinholeCamera: Equatable, Sendable {
    var fx: Double
    var fy: Double
    var cx: Double
    var cy: Double
    var width: Int
    var height: Int
}

/// Clockwise rotation in quarter-turns. The raw value is transmitted as `image.orientation_applied`.
enum QuarterTurn: Int, Sendable, CaseIterable {
    case none = 0
    case cw90 = 1
    case cw180 = 2
    case cw270 = 3

    /// Rotation needed to bring a sensor-native landscape frame upright for the given device
    /// orientation. `nil` for face-up/face-down/unknown, where the caller should reuse the last
    /// known good value rather than guessing.
    static func uprightRotation(forDeviceOrientation orientation: DeviceOrientationInput) -> QuarterTurn? {
        switch orientation {
        case .portrait: .cw90
        case .portraitUpsideDown: .cw270
        case .landscapeLeft: .none
        case .landscapeRight: .cw180
        case .unknown: nil
        }
    }
}

/// Deliberately not `UIDeviceOrientation`, so the transform layer stays free of UIKit and
/// remains testable on any platform.
enum DeviceOrientationInput: Sendable {
    case portrait
    case portraitUpsideDown
    case landscapeLeft
    case landscapeRight
    case unknown
}

/// Rotation and resampling of an image *together with* its intrinsics and the worker's tap.
///
/// Every function here is pure and total. The whole point of this type is that the image, K,
/// and the target pixel are transformed in one call and cannot drift apart; see
/// `docs/INTRINSICS.md` section 2 for the derivations these implement.
enum ImageTransform {

    /// Longest side of the transmitted image. The 2026_07_08 package advises <= 1280 px for the
    /// MultiSet query, and K is derived from what is actually sent, never from the original.
    static let maxUploadLongSide = 1280

    // MARK: - Rotation

    /// Size of the frame after `turn` is applied.
    static func rotatedSize(width: Int, height: Int, by turn: QuarterTurn) -> (width: Int, height: Int) {
        switch turn {
        case .none, .cw180: (width, height)
        case .cw90, .cw270: (height, width)
        }
    }

    /// Rotate intrinsics by `turn` quarter-turns clockwise.
    ///
    /// | k | size    | fx' | fy' | cx'    | cy'    |
    /// |---|---------|-----|-----|--------|--------|
    /// | 1 | H x W   | fy  | fx  | H - cy | cx     |
    /// | 2 | W x H   | fx  | fy  | W - cx | H - cy |
    /// | 3 | H x W   | fy  | fx  | cy     | W - cx |
    static func rotate(_ camera: PinholeCamera, by turn: QuarterTurn) -> PinholeCamera {
        let w = Double(camera.width)
        let h = Double(camera.height)
        let size = rotatedSize(width: camera.width, height: camera.height, by: turn)

        switch turn {
        case .none:
            return camera
        case .cw90:
            return PinholeCamera(fx: camera.fy, fy: camera.fx,
                                 cx: h - camera.cy, cy: camera.cx,
                                 width: size.width, height: size.height)
        case .cw180:
            return PinholeCamera(fx: camera.fx, fy: camera.fy,
                                 cx: w - camera.cx, cy: h - camera.cy,
                                 width: size.width, height: size.height)
        case .cw270:
            return PinholeCamera(fx: camera.fy, fy: camera.fx,
                                 cx: camera.cy, cy: w - camera.cx,
                                 width: size.width, height: size.height)
        }
    }

    /// Rotate a pixel coordinate by the same `turn`, in a frame of size `width` x `height`.
    static func rotate(point: CGPoint, in width: Int, _ height: Int, by turn: QuarterTurn) -> CGPoint {
        let w = Double(width)
        let h = Double(height)
        switch turn {
        case .none: return point
        case .cw90: return CGPoint(x: h - point.y, y: point.x)
        case .cw180: return CGPoint(x: w - point.x, y: h - point.y)
        case .cw270: return CGPoint(x: point.y, y: w - point.x)
        }
    }

    // MARK: - Scaling

    /// Output size that puts the longest side at `longSide`, preserving aspect ratio.
    /// Never upscales - an image smaller than the target is transmitted untouched.
    static func downscaledSize(width: Int, height: Int, longSide: Int = maxUploadLongSide) -> (width: Int, height: Int) {
        let longest = max(width, height)
        guard longest > longSide else { return (width, height) }
        let factor = Double(longSide) / Double(longest)
        return (max(1, Int((Double(width) * factor).rounded())),
                max(1, Int((Double(height) * factor).rounded())))
    }

    /// Scale intrinsics to a new frame size.
    ///
    /// Uses the realised per-axis ratios `outW/inW` and `outH/inH` rather than the requested
    /// factor: rounding the output dimensions to integers makes the two axes differ slightly,
    /// and each of fx, fy, cx, cy is linear in the resolution of the axis it belongs to.
    static func scale(_ camera: PinholeCamera, toWidth outWidth: Int, height outHeight: Int) -> PinholeCamera {
        let sx = Double(outWidth) / Double(camera.width)
        let sy = Double(outHeight) / Double(camera.height)
        return PinholeCamera(fx: camera.fx * sx, fy: camera.fy * sy,
                             cx: camera.cx * sx, cy: camera.cy * sy,
                             width: outWidth, height: outHeight)
    }

    /// Scale a pixel coordinate by the same per-axis ratios.
    static func scale(point: CGPoint, from input: (width: Int, height: Int), to output: (width: Int, height: Int)) -> CGPoint {
        CGPoint(x: point.x * Double(output.width) / Double(input.width),
                y: point.y * Double(output.height) / Double(input.height))
    }

    // MARK: - The composed pipeline

    struct Result: Equatable, Sendable {
        var camera: PinholeCamera
        var target: CGPoint
        var turn: QuarterTurn
        var sourceWidth: Int
        var sourceHeight: Int
        /// Realised uniform scale, for provenance. K is already scaled.
        var scale: Double
    }

    /// Apply `rotate` then `downscale` to intrinsics and the tap point together.
    ///
    /// The caller is responsible for pushing the same `turn` and the same output size through
    /// the pixel pipeline - `PixelBufferRenderer` takes this `Result` so the two cannot diverge.
    static func apply(
        camera: PinholeCamera,
        target: CGPoint,
        turn: QuarterTurn,
        longSide: Int = maxUploadLongSide
    ) -> Result {
        let sourceWidth = camera.width
        let sourceHeight = camera.height

        let rotatedCamera = rotate(camera, by: turn)
        let rotatedTarget = rotate(point: target, in: sourceWidth, sourceHeight, by: turn)

        let outSize = downscaledSize(width: rotatedCamera.width, height: rotatedCamera.height, longSide: longSide)
        let finalCamera = scale(rotatedCamera, toWidth: outSize.width, height: outSize.height)
        let finalTarget = scale(point: rotatedTarget,
                                from: (rotatedCamera.width, rotatedCamera.height),
                                to: outSize)

        return Result(
            camera: finalCamera,
            target: clamp(finalTarget, toWidth: outSize.width, height: outSize.height),
            turn: turn,
            sourceWidth: sourceWidth,
            sourceHeight: sourceHeight,
            scale: Double(outSize.width) / Double(rotatedCamera.width)
        )
    }

    /// Keep a tap strictly inside the frame. A tap exactly on the right or bottom edge would
    /// otherwise produce an out-of-range pixel index downstream.
    static func clamp(_ point: CGPoint, toWidth width: Int, height: Int) -> CGPoint {
        CGPoint(x: min(max(point.x, 0), Double(width) - 1),
                y: min(max(point.y, 0), Double(height) - 1))
    }

    /// 0 at the frame centre, 1 at the edge. Section 6.2 of the calibration note uses this to
    /// reject captures where lens distortion makes the pinhole model untenable.
    static func centrality(of point: CGPoint, inWidth width: Int, height: Int) -> Double {
        guard width > 0, height > 0 else { return 1 }
        let dx = abs(point.x / Double(width) - 0.5)
        let dy = abs(point.y / Double(height) - 0.5)
        return min(1, max(dx, dy) / 0.5)
    }

    /// Above this the worker is asked to re-frame with the defect nearer the centre.
    static let centralityWarningThreshold = 0.6
}

File: android/shared/src/commonMain/kotlin/ai/cbm/capture/domain/imaging/ImageTransform.kt
package ai.cbm.capture.domain.imaging

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * A pinhole camera in the pixel frame of a specific image.
 *
 * The [width]/[height] fields are what make this type meaningful: an intrinsics matrix without
 * the frame it belongs to is exactly the defect described in section 3 of the calibration note,
 * so the two never travel separately here.
 */
data class PinholeCamera(
    val fx: Double,
    val fy: Double,
    val cx: Double,
    val cy: Double,
    val width: Int,
    val height: Int
)

data class PointF2(val x: Double, val y: Double)

/** Clockwise rotation in quarter-turns; the ordinal is transmitted as `image.orientation_applied`. */
enum class QuarterTurn(val turns: Int) {
    NONE(0), CW90(1), CW180(2), CW270(3);

    companion object {
        /**
         * Rotation that brings a sensor-native landscape frame upright, given the display
         * rotation reported by the window manager (`Surface.ROTATION_*`).
         */
        fun forDisplayRotation(surfaceRotation: Int): QuarterTurn = when (surfaceRotation) {
            0 -> CW90     // Surface.ROTATION_0, portrait
            1 -> NONE     // Surface.ROTATION_90
            2 -> CW270    // Surface.ROTATION_180
            3 -> CW180    // Surface.ROTATION_270
            else -> CW90
        }
    }
}

/**
 * Rotation and resampling of an image *together with* its intrinsics and the worker's tap.
 *
 * Every function here is pure and total, and each is the Kotlin twin of the Swift implementation
 * in `ios/CBMCapture/Core/Imaging/ImageTransform.swift`. Both are implementations of section 2
 * of `docs/INTRINSICS.md`, and both test suites assert the same identities, so the two platforms
 * cannot drift apart silently.
 */
object ImageTransform {

    /**
     * Longest side of the transmitted image. The 2026_07_08 package advises <= 1280 px for the
     * MultiSet query, and K is derived from what is actually sent, never from the original.
     */
    const val MAX_UPLOAD_LONG_SIDE = 1280

    /** Above this, the worker is asked to re-frame with the defect nearer the centre. */
    const val CENTRALITY_WARNING_THRESHOLD = 0.6

    // ---- Rotation ----

    fun rotatedSize(width: Int, height: Int, turn: QuarterTurn): Pair<Int, Int> =
        when (turn) {
            QuarterTurn.NONE, QuarterTurn.CW180 -> width to height
            QuarterTurn.CW90, QuarterTurn.CW270 -> height to width
        }

    /**
     * Rotate intrinsics by [turn] quarter-turns clockwise.
     *
     * | k | size  | fx' | fy' | cx'    | cy'    |
     * |---|-------|-----|-----|--------|--------|
     * | 1 | H x W | fy  | fx  | H - cy | cx     |
     * | 2 | W x H | fx  | fy  | W - cx | H - cy |
     * | 3 | H x W | fy  | fx  | cy     | W - cx |
     */
    fun rotate(camera: PinholeCamera, turn: QuarterTurn): PinholeCamera {
        val w = camera.width.toDouble()
        val h = camera.height.toDouble()
        val (outW, outH) = rotatedSize(camera.width, camera.height, turn)

        return when (turn) {
            QuarterTurn.NONE -> camera
            QuarterTurn.CW90 -> PinholeCamera(camera.fy, camera.fx, h - camera.cy, camera.cx, outW, outH)
            QuarterTurn.CW180 -> PinholeCamera(camera.fx, camera.fy, w - camera.cx, h - camera.cy, outW, outH)
            QuarterTurn.CW270 -> PinholeCamera(camera.fy, camera.fx, camera.cy, w - camera.cx, outW, outH)
        }
    }

    fun rotate(point: PointF2, width: Int, height: Int, turn: QuarterTurn): PointF2 {
        val w = width.toDouble()
        val h = height.toDouble()
        return when (turn) {
            QuarterTurn.NONE -> point
            QuarterTurn.CW90 -> PointF2(h - point.y, point.x)
            QuarterTurn.CW180 -> PointF2(w - point.x, h - point.y)
            QuarterTurn.CW270 -> PointF2(point.y, w - point.x)
        }
    }

    // ---- Scaling ----

    /** Never upscales: an image already below the limit is transmitted untouched. */
    fun downscaledSize(width: Int, height: Int, longSide: Int = MAX_UPLOAD_LONG_SIDE): Pair<Int, Int> {
        val longest = max(width, height)
        if (longest <= longSide) return width to height
        val factor = longSide.toDouble() / longest
        return max(1, (width * factor).roundToInt()) to max(1, (height * factor).roundToInt())
    }

    /**
     * Scale intrinsics to a new frame size using the **realised** per-axis ratios.
     *
     * Rounding the output dimensions to integers makes the two axes differ slightly, and each of
     * fx, fy, cx, cy is linear in the resolution of the axis it belongs to. Applying one factor
     * to all four, or scaling only the focal lengths, is the classic form of this bug.
     */
    fun scale(camera: PinholeCamera, outWidth: Int, outHeight: Int): PinholeCamera {
        val sx = outWidth.toDouble() / camera.width
        val sy = outHeight.toDouble() / camera.height
        return PinholeCamera(camera.fx * sx, camera.fy * sy, camera.cx * sx, camera.cy * sy, outWidth, outHeight)
    }

    fun scale(point: PointF2, inWidth: Int, inHeight: Int, outWidth: Int, outHeight: Int): PointF2 =
        PointF2(point.x * outWidth / inWidth, point.y * outHeight / inHeight)

    // ---- The composed pipeline ----

    data class Result(
        val camera: PinholeCamera,
        val target: PointF2,
        val turn: QuarterTurn,
        val sourceWidth: Int,
        val sourceHeight: Int,
        /** Realised uniform scale, for provenance. K is already scaled. */
        val scale: Double
    )

    /**
     * Apply rotation then downscale to intrinsics and the tap point together.
     *
     * The caller pushes the returned [Result] through the pixel pipeline as well, so the pixels,
     * K, and the tap cannot end up in different coordinate systems.
     */
    fun apply(
        camera: PinholeCamera,
        target: PointF2,
        turn: QuarterTurn,
        longSide: Int = MAX_UPLOAD_LONG_SIDE
    ): Result {
        val sourceWidth = camera.width
        val sourceHeight = camera.height

        val rotatedCamera = rotate(camera, turn)
        val rotatedTarget = rotate(target, sourceWidth, sourceHeight, turn)

        val (outW, outH) = downscaledSize(rotatedCamera.width, rotatedCamera.height, longSide)
        val finalCamera = scale(rotatedCamera, outW, outH)
        val finalTarget = scale(rotatedTarget, rotatedCamera.width, rotatedCamera.height, outW, outH)

        return Result(
            camera = finalCamera,
            target = clamp(finalTarget, outW, outH),
            turn = turn,
            sourceWidth = sourceWidth,
            sourceHeight = sourceHeight,
            scale = outW.toDouble() / rotatedCamera.width
        )
    }

    /** A tap exactly on the right or bottom edge would produce an out-of-range pixel index. */
    fun clamp(point: PointF2, width: Int, height: Int): PointF2 =
        PointF2(
            min(max(point.x, 0.0), width - 1.0),
            min(max(point.y, 0.0), height - 1.0)
        )

    /**
     * 0 at the frame centre, 1 at the edge. Section 6.2 of the calibration note uses this to
     * reject captures where lens distortion makes the pinhole model untenable.
     */
    fun centrality(point: PointF2, width: Int, height: Int): Double {
        if (width <= 0 || height <= 0) return 1.0
        val dx = abs(point.x / width - 0.5)
        val dy = abs(point.y / height - 0.5)
        return min(1.0, max(dx, dy) / 0.5)
    }
}
